lib/galaxy/tools/toolbox/base.py

```diff
@@ -927,53 +927,59 @@ def tool_panel_contents(self, trans, **kwds):
+        Note: The cached tool's to_dict is specific to the calls from toolbox.
+        """
+        if not trans.user_is_admin():
+            self._tool_to_dict_cache[tool_id] = to_dict
+        else:
+            self._tool_to_dict_cache_admin[tool_id] = to_dict
+
     def to_dict(self, trans, in_panel=True, **kwds):
         """
         Create a dictionary representation of the toolbox.
-        Uses primitive cache for Galaxies without filters.
+        Uses primitive cache for toolbox-specific tool 'to_dict's.
         """
-        filters_enabled = self.check_for_filters(trans)
         if in_panel:
-            if not filters_enabled and self._cached_to_dict_panel:
-                return self._cached_to_dict_panel
-            else:
-                panel_elts = list(self.tool_panel_contents(trans, **kwds))
-                rval = []
-                kwargs = dict(
-                    trans=trans,
-                    link_details=True
-                )
-                for elt in panel_elts:
-                    rval.append(elt.to_dict(**kwargs))
-                if not filters_enabled:
-                    self._cached_to_dict_panel = rval
+            panel_elts = list(self.tool_panel_contents(trans, **kwds))
+            rval = []
+            for elt in panel_elts:
+                cached = self._get_tool_to_dict_from_cache(trans, elt.id)
+                if cached:
+                    rval.append(cached)
+                else:
+                    elt_to_dict = elt.to_dict(trans=trans, link_details=True)
+                    rval.append(elt_to_dict)
+                    self._store_tool_to_dict_in_cache(trans, elt.id, elt_to_dict)
         else:
-            if not filters_enabled and self._cached_to_dict_all:
-                return self._cached_to_dict_all
-            else:
-                filter_method = self._build_filter_method(trans)
-                tools = []
-                for id, tool in self._tools_by_id.items():
-                    tool = filter_method(tool, panel_item_types.TOOL)
-                    if not tool:
-                        continue
-                    tools.append(tool.to_dict(trans, link_details=True))
-                rval = tools
-                if not filters_enabled:
-                    self._cached_to_dict_all = rval
+            filter_method = self._build_filter_method(trans)
+            tools = []
+            for id, tool in self._tools_by_id.items():
+                tool = filter_method(tool, panel_item_types.TOOL)
+                if not tool:
+                    continue
+                cached = self._get_tool_to_dict_from_cache(trans, tool.id)
+                if cached:
+                    tools.append(cached)
+                else:
+                    tool_to_dict = tool.to_dict(trans, link_details=True)
+                    tools.append(tool_to_dict)
+                    self._store_tool_to_dict_in_cache(trans, tool.id, tool_to_dict)
+            rval = tools
```

Comment:
Can you just use `rval` instead of `tools` everywhere in lines 968-978 and remove this line?
In that case you could also move the common `rval = []` initialisation before the `if`.